### PR TITLE
Support Parquet Benchmark for BigInt and Double with different Filter and Null rates

### PR DESCRIPTION
**Pull request by liushengxuan**
This PR supports Parquet Benchmark for BigInt and Double with different Filter and Null rates on both Dictionary and Plain encoded Parquet files. Including:

1. Refactored the initial parquet benchmark so that it is able to support the benchmark on different types, filter rates, null rates and both Dictionary/Plain Parquet file encoding types.
2. Added the the element counter to the filter generator and increased the sample size to 65536 for more accurate filter rate estimation. Also, Added `allowNulls` to `FilterSpec` to generator non-null filters for the benchmark.

### CODE REVIEW

File: velox/dwio/common/tests/utils/FilterGenerator.h [hunk 1]
```diff
@@ -71,8 +71,8 @@ class AbstractColumnStats {
   // ASCII string greater than test data values. Used for row group skipping
   // tests.
   static constexpr const char* kMaxString = "~~~~~";
-  AbstractColumnStats(TypePtr type, RowTypePtr rootType)
-      : type_(type), rootType_(rootType) {}
+  AbstractColumnStats(TypePtr type, RowTypePtr rootType, bool allowNulls = true)
```
Comment (yingsu00) [inline velox/dwio/common/tests/utils/FilterGenerator.h:74]: allowNulls is a property of filters, not ColumnStats. I think it would be cleaner to add allowNulls to FilterSpec, and pass it in through ColumnStats::filter(FilterSpec& spec) - > makeRangeFilter(spec)

File: velox/dwio/common/tests/utils/FilterGenerator.h [hunk 2]
```diff
@@ -242,12 +245,19 @@ class ColumnStats : public AbstractColumnStats {
...
+        return std::make_unique<velox::common::IsNull>();
+      } else {
+        return std::make_unique<velox::common::BigintRange>(0, 0, false);
+      }
     }
     int32_t lowerIndex;
     int32_t upperIndex;
     T lower = valueAtPct(startPct, &lowerIndex);
     T upper = valueAtPct(startPct + selectPct, &upperIndex);
+    if (!allowNulls_) {
```
Comment (yingsu00) [inline velox/dwio/common/tests/utils/FilterGenerator.h:258]: Using allowNulls_ to detect if the caller wants to just get this range filter seems a bit hacky. Currently only the ParquetReaderBenchmark set it to false, but in the future other tests may set it to false. Then this function would never run the below lines. If we pass it from FilterSpec, the callers have more flexibility.

**Comment by yingsu00** [comment]
@liushengxuan Shengxuan, thank you very much for improving the benchmark! Could you include the benchmark result at the end of ParquetReaderBenchmark.cpp? You can use #ifdef 0 to comment it out. Also include the CPU model.

Another catch was, the data was actually dictionary encoded, not plain. Would you please rename `readSingleColumnPlain` to just `readSingleColumn`, and add benchmarks for plain encoding too? To do that you need to set the writer property
```
writerProperties_ = ::parquet::WriterProperties::Builder()
                          .disable_dictionary()
                          ->build();
```

File: velox/dwio/parquet/tests/reader/ParquetReaderBenchmark.cpp [hunk 3]
```diff
@@ -33,11 +33,10 @@ using namespace facebook::velox::dwio::common;
 using namespace facebook::velox::parquet;
 using namespace facebook::velox::test;
 
-// TODO: Note that the kNumRowsPerBatch needs to be a small number for now. For
-// details please see https://github.com/facebookincubator/velox/issues/2844
 const uint32_t kNumRowsPerBatch = 60000;
 const uint32_t kNumBatches = 40;
 const uint32_t kNumRowsPerRowGroup = 10000;
+const double kExpectedThreshold = 0.5;
```
[the remark this answers is not in the document]
Comment (liushengxuan) [reply velox/dwio/parquet/tests/reader/ParquetReaderBenchmark.cpp:39]: SInce the sample size is 4096, there are some edge cases that the expected is very different from the actual size. I can try to make the 4096 larger

File: velox/dwio/common/tests/utils/FilterGenerator.h [hunk 4]
```diff
@@ -242,12 +245,19 @@ class ColumnStats : public AbstractColumnStats {
 
   std::unique_ptr<Filter> makeRangeFilter(float startPct, float selectPct) {
     if (values_.empty()) {
-      return std::make_unique<velox::common::IsNull>();
+      if (allowNulls_) {
```
[the remark this answers is not in the document]
Comment (liushengxuan) [reply velox/dwio/common/tests/utils/FilterGenerator.h:248]: Yes. Originally, the it generates a `IsNull` filter, which returns `true` on nulls. But this time, we need to return `false` on nulls

**Comment by yingsu00** [comment]
@liushengxuan Running it in release mode gives this error:

E20221116 21:24:52.657351 20850011 Exceptions.h:68] Line: /Users/yingsu/repo/velox1/velox/velox/dwio/parquet/tests/reader/ParquetReaderBenchmark.cpp:202, Function:readSingleColumnPlain, Expression: resultSize <= upperBound && resultSize >= lowerBound Result Size and Expected Size Mismatch, Source: RUNTIME, ErrorCode: INVALID_STATE

**Comment by liushengxuan** [comment]
> @liushengxuan Running it in release mode gives this error:
> 
> E20221116 21:24:52.657351 20850011 Exceptions.h:68] Line: /Users/yingsu/repo/velox1/velox/velox/dwio/parquet/tests/reader/ParquetReaderBenchmark.cpp:202, Function:readSingleColumnPlain, Expression: resultSize <= upperBound && resultSize >= lowerBound Result Size and Expected Size Mismatch, Source: RUNTIME, ErrorCode: INVALID_STATE

Yes. It is a known issue. This is caused by the current sampling, which takes 4096 unique elements, without considering the number of appearance. So there is a chance that the filter error rate is pretty hight. I am working on this issue

**Comment by liushengxuan** [comment]
> @liushengxuan Running it in release mode gives this error:
> 
> E20221116 21:24:52.657351 20850011 Exceptions.h:68] Line: /Users/yingsu/repo/velox1/velox/velox/dwio/parquet/tests/reader/ParquetReaderBenchmark.cpp:202, Function:readSingleColumnPlain, Expression: resultSize <= upperBound && resultSize >= lowerBound Result Size and Expected Size Mismatch, Source: RUNTIME, ErrorCode: INVALID_STATE

I believe this issue is fixed in the PR. I introduced a counter and increased the sample size. So the estimated filter rate is more accurate.

**Comment by liushengxuan** [comment]
> @liushengxuan Shengxuan, thank you very much for improving the benchmark! Could you include the benchmark result at the end of ParquetReaderBenchmark.cpp? You can use #ifdef 0 to comment it out. Also include the CPU model.
> 
> Another catch was, the data was actually dictionary encoded, not plain. Would you please rename `readSingleColumnPlain` to just `readSingleColumn`, and add benchmarks for plain encoding too? To do that you need to set the writer property
> 
> ```
> writerProperties_ = ::parquet::WriterProperties::Builder()
>                           .disable_dictionary()
>                           ->build();
> ```

Addressed

File: velox/dwio/common/tests/utils/FilterGenerator.cpp [hunk 5]
```diff
@@ -91,7 +92,8 @@ std::unique_ptr<Filter> ColumnStats<bool>::makeRangeFilter(
 template <>
 std::unique_ptr<Filter> ColumnStats<UnscaledShortDecimal>::makeRangeFilter(
     float startPct,
-    float selectPct) {
+    float selectPct,
+    FilterSpec filterSpec) {
```
Comment (yingsu00) [inline velox/dwio/common/tests/utils/FilterGenerator.cpp:96]: FilterSpec contains startPct and selectPct already. It might be better to create two set of signatures
```
template <>
std::unique_ptr<Filter> ColumnStats<>::makeRangeFilter(FilterSpec filterSpec) {
...
}

template <typename T>
std::unique_ptr<Filter> ColumnStats<>::makeRangeFilter(float startPct, float selectPct) {
    FilterSpec filterSpec(startPct, selectPct);
    return makeRangeFilter<T>(filterSpec);
}
```

File: velox/dwio/common/tests/utils/FilterGenerator.cpp [hunk 6]
```diff
@@ -147,17 +150,34 @@ std::unique_ptr<Filter> ColumnStats<float>::makeRangeFilter(
...
 std::unique_ptr<Filter> ColumnStats<double>::makeRangeFilter(
     float startPct,
-    float selectPct) {
+    float selectPct,
+    FilterSpec filterSpec) {
   if (values_.empty()) {
-    return std::make_unique<velox::common::IsNull>();
+    if (filterSpec.allowNulls_) {
+      return std::make_unique<velox::common::IsNull>();
+    } else {
```
Comment (yingsu00) [inline velox/dwio/common/tests/utils/FilterGenerator.cpp:158]: Does it matter to return different filters according to allowNulls_ when the values are empty? It looks unnecessary for me.

File: velox/dwio/common/tests/utils/FilterGenerator.h [hunk 7]
```diff
@@ -86,7 +89,7 @@ class AbstractColumnStats {
       float selectPct,
       FilterKind filterKind,
       const std::vector<RowVectorPtr>& batches,
-      const Subfield& subfield,
+      const FilterSpec& filterSpec,
```
Comment (yingsu00) [inline velox/dwio/common/tests/utils/FilterGenerator.h:92]: Same, FilterSpec already contains startPct, selectPct and filterKind. Other places too.

File: velox/dwio/common/tests/utils/FilterGenerator.cpp (hunk 6 above)
Comment (liushengxuan) [reply velox/dwio/common/tests/utils/FilterGenerator.cpp:158]: Well. It's `IsNull` filter now, which returns true on nulls. But it's true that I can merge them

File: CMakeLists.txt [hunk 8]
```diff
@@ -64,7 +64,6 @@ option(VELOX_ENABLE_HDFS "Build Hdfs Connector" OFF)
 option(VELOX_ENABLE_PARQUET "Enable Parquet support" OFF)
 option(VELOX_ENABLE_ARROW "Enable Arrow support" OFF)
 option(VELOX_BUILD_TEST_UTILS "Enable Velox test utilities" OFF)
-option(VELOX_BUILD_PYTHON_PACKAGE "Enable Velox Python bindings" OFF)
```
[the remark this answers is not in the document]
Comment (liushengxuan) [reply CMakeLists.txt:67]: Oh boy good catch. Thanks

File: velox/dwio/common/tests/utils/FilterGenerator.cpp [hunk 9]
```diff
@@ -141,46 +142,57 @@ std::unique_ptr<Filter> ColumnStats<float>::makeRangeFilter(
...
 }
 
 template <>
 std::unique_ptr<Filter> ColumnStats<double>::makeRangeFilter(
-    float startPct,
-    float selectPct) {
+    FilterSpec filterSpec) {
   if (values_.empty()) {
-    return std::make_unique<velox::common::IsNull>();
+    return std::make_unique<velox::common::DoubleRange>(
```
Comment (yingsu00) [inline velox/dwio/common/tests/utils/FilterGenerator.cpp:152]: Was there any reason you changed the filter for empty values for this double? For other types there are sometimes checking on allowNulls_, sometimes not. If there are no specific reasons, let's revert the changes. In my understanding, the allowNulls_ doesn't matter for empty values.

Comment (liushengxuan) [reply velox/dwio/common/tests/utils/FilterGenerator.cpp:152]: I see your point. Before introducing this PR, the double filter returns true all the nulls. And also, if all the elements in the vector are nulls, `values_.empty()` returns true. In the current PR, we need it to return false on all the nulls, so we need it here.

**Review by yingsu00 — APPROVED** [review]

**Comment by liushengxuan** [comment]
@Yuhta this PR is the same as https://github.com/facebookincubator/velox/pull/3353

I will fix the #if 0 issue later today for both. Thanks!

**Comment by liushengxuan** [comment]
> @Yuhta this PR is the same as #3353
> 
> I will fix the #if 0 issue later today for both. Thanks!

@Yuhta I have fixed the #if 0 comment issue. Can you please review it? Thanks!!

File: velox/dwio/parquet/tests/reader/ParquetReaderBenchmark.cpp [hunk 10]
```diff
@@ -64,13 +72,44 @@ class ParquetReaderBenchmark {
     writer_->flush();
   }
 
+  FilterSpec createFilterSpec(
+      const std::string columnName,
```
Comment (Yuhta) [inline velox/dwio/parquet/tests/reader/ParquetReaderBenchmark.cpp:76]: `const std::string&`

File: velox/dwio/parquet/tests/reader/ParquetReaderBenchmark.cpp [hunk 11]
```diff
@@ -33,20 +33,28 @@ using namespace facebook::velox::dwio::common;
...
 const uint32_t kNumRowsPerBatch = 60000;
-const uint32_t kNumBatches = 40;
+const uint32_t kNumBatches = 50;
 const uint32_t kNumRowsPerRowGroup = 10000;
+const double kFilterErrorMargin = 0.2;
 
 class ParquetReaderBenchmark {
  public:
-  ParquetReaderBenchmark() {
+  ParquetReaderBenchmark(bool disableDictionary)
```
Comment (Yuhta) [inline velox/dwio/parquet/tests/reader/ParquetReaderBenchmark.cpp:43]: `explicit` for all constructor taking one parameter

**Comment by Yuhta** [comment]
@liushengxuan Also can you run dwrf `E2EFilterTest` under ASAN?  There are some undefined behavior with changes in `FilterGenerator`.
